source/server/hot_restarting_parent.cc

```diff
@@ -107,10 +109,13 @@ HotRestartingParent::Internal::getListenSocketsForChild(const HotRestartMessage:
   for (const auto& listener : server_->listenerManager().listeners()) {
     Network::ListenSocketFactory& socket_factory = listener.get().listenSocketFactory();
     if (*socket_factory.localAddress() == *addr && listener.get().bindToPort()) {
-      if (socket_factory.sharedSocket().has_value()) {
-        // Pass the socket to the new process if it is already shared across workers.
+      // worker_index() will default to 0 if not set which is the behavior before this field
+      // was added. Thus, this should be safe for both roll forward and roll back.
+      if (request.pass_listen_socket().worker_index() < server_->options().concurrency()) {
         wrapped_reply.mutable_reply()->mutable_pass_listen_socket()->set_fd(
-            socket_factory.sharedSocket()->get().ioHandle().fdDoNotUse());
+            socket_factory.getListenSocket(request.pass_listen_socket().worker_index())
+                ->ioHandle()
+                .fdDoNotUse());
```

Comment:
Nit: Should we kill the fdDoNotuse? I remember IoHandle::duplicate is to resolve `do not use`

Reply:
Passing the socket requires the actual fd number.